cmd/infracost/run.go

```diff
@@ -802,14 +808,95 @@ func AddHCLEnvVars(projectContexts []*config.ProjectContext, r output.Root, hclR
 	env["tfRunTimeMs"] = tfTimeTaken
 	env["hclRunTimeMs"] = hclTimeTaken
 
+	diff := collectHCLRunDiff(projects, hclProjects)
+	env["hclMissingResources"] = diff.missingResources
+	env["hclResourceDiff"] = diff.resourceDiffs
+}
+
+func collectHCLRunDiff(projects, hclProjects []*schema.Project) hclRunDiff {
+	diff := map[string][]string{}
+	missingResources := map[string]bool{}
+
+	hclProjectsMapping := map[string]*schema.Project{}
+	for _, project := range hclProjects {
+		hclProjectsMapping[project.Name] = project
+	}
+
+	for _, project := range projects {
+		hclProject := hclProjectsMapping[project.Name]
+
+		if hclProject == nil {
+			log.Debugf("could not find a matching HCL project '%s' for HCL run diff", project.Name)
+			continue
+		}
+
+		hclResourcesMapping := map[string]*decimal.Decimal{}
+		hclResourceNames := map[string]bool{}
```

Comment:
this lookup map is not needed any more - you can just assert that the `hclResourcesMapping`, e.g.

```go
	if _, ok := hclResourcesMapping[resource.Name]; !ok {
		missingResources[resource.ResourceType] = true
                continue
	}
```